File: app/src/main/java/com/example/diffutilrv/veiwmodel/MainViewModel.java
package com.example.diffutilrv.veiwmodel;

import androidx.hilt.lifecycle.ViewModelInject;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Transformations;
import androidx.lifecycle.ViewModel;

import com.example.diffutilrv.model.Employee;
import com.example.diffutilrv.usecase.EmployeeUseCase;

import java.util.List;

public class MainViewModel extends ViewModel {

    private final EmployeeUseCase mEmployeeUseCase;
    private final MutableLiveData<EmployeeSortType> mEmployeeSortTypeLiveData;
    private LiveData<List<Employee>> mEmployeeListLiveData;

    @ViewModelInject
    public MainViewModel(EmployeeUseCase employeeUseCase) {
        this.mEmployeeUseCase = employeeUseCase;
        this.mEmployeeSortTypeLiveData = new MutableLiveData<>(EmployeeSortType.ROLE);
    }

    public LiveData<List<Employee>> getEmployeeListLiveData() {
        if (mEmployeeListLiveData == null) {
            mEmployeeListLiveData = Transformations.map(mEmployeeSortTypeLiveData, employeeSortType -> {
                switch (employeeSortType) {
                    case NAME:
                        return mEmployeeUseCase.getEmployeeListSortedByName();
                    case ROLE:
                    default:
                        return mEmployeeUseCase.getEmployeeListSortedByRole();
                }
            });
        }
        return mEmployeeListLiveData;
    }

    public void upDataSortType(EmployeeSortType employeeSortType) {
        this.mEmployeeSortTypeLiveData.postValue(employeeSortType);
    }
}

File: app/src/main/java/com/example/diffutilrv/MainViewModel.kt
package com.example.diffutilrv

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import dagger.hilt.android.lifecycle.HiltViewModel
import javax.inject.Inject

private const val STATE_SORT_BY = "sortBy"

@HiltViewModel
class MainViewModel @Inject constructor(
    private val state: SavedStateHandle,
    private val dataUtils: DummyEmployeeDataUtils,
) : ViewModel() {
    private val _employees = MutableLiveData<List<Employee>>(emptyList())
    val employees: LiveData<List<Employee>> = _employees
    private val _clickRowAction = MutableLiveData<OneOffEvent<Employee>>()
    val clickRowAction: LiveData<OneOffEvent<Employee>> = _clickRowAction
    private val _clickRowButtonAction = MutableLiveData<OneOffEvent<Employee>>()
    val clickRowButtonAction: LiveData<OneOffEvent<Employee>> = _clickRowButtonAction

    fun loadData(sortBy: EmployeeSortBy? = null) {
        _employees.value = when (sortBy ?: state[STATE_SORT_BY] ?: EmployeeSortBy.ROLE) {
            EmployeeSortBy.ROLE -> dataUtils.getEmployeeListSortedByRole()
            EmployeeSortBy.NAME -> dataUtils.getEmployeeListSortedByName()
        }
    }

    fun changeSorting(sortBy: EmployeeSortBy) {
        state[STATE_SORT_BY] = sortBy
        loadData(sortBy)
    }

    fun onClickRow(employee: Employee) {
        _clickRowAction.value = OneOffEvent(employee)
    }

    fun onClickRowButton(employee: Employee) {
        _clickRowButtonAction.value = OneOffEvent(employee)
    }
}